src/webgpu/util/copyToTexture.ts

```diff
@@ -0,0 +1,126 @@
+export const description = `
+copyToTexture with HTMLCanvasElement and OffscreenCanvas sources.
+
+- x= {HTMLCanvasElement, OffscreenCanvas}
+- x= {2d, webgl, webgl2, gpupresent, bitmaprenderer} context, {various context creation attributes}
+
+TODO: consider whether external_texture and copyToTexture video tests should be in the same file
+TODO: plan
+`;
+
+import { GPUTest } from '../gpu_test.js';
+
+function calculateRowPitch(width: number, bytesPerPixel: number): number {
+  const bytesPerRow = width * bytesPerPixel;
+  // Rounds up to a multiple of 256 according to WebGPU requirements.
+  return (((bytesPerRow - 1) >> 8) + 1) << 8;
+}
+
+export class CopyToTextureUtils extends GPUTest {
+  checkCopyExternalImageResult(
+    src: GPUBuffer,
+    expected: ArrayBufferView,
+    width: number,
+    height: number,
+    bytesPerPixel: number
+  ): void {
+    const exp = new Uint8Array(expected.buffer, expected.byteOffset, expected.byteLength);
+    const rowPitch = calculateRowPitch(width, bytesPerPixel);
+    const dst = this.createCopyForMapRead(src, 0, rowPitch * height);
+
+    this.eventualAsyncExpectation(async niceStack => {
+      await dst.mapAsync(GPUMapMode.READ);
+      const actual = new Uint8Array(dst.getMappedRange());
+      const check = this.checkBufferWithRowPitch(
+        actual,
+        exp,
+        width,
+        height,
+        rowPitch,
+        bytesPerPixel
+      );
+      if (check !== undefined) {
+        niceStack.message = check;
+        this.rec.expectationFailed(niceStack);
+      }
+      dst.destroy();
+    });
+  }
+
+  checkBufferWithRowPitch(
```

Comment:
This duplicates a lot with the original `checkBuffer`, can we simplify this by doing something like splitting the check row by row, and returning the first error?

```ts
for (let y = 0; y < height; ++y) {
  const checkResult = this.checkBuffer(just the current row);
  if (checkResult !== undefined) return `on row ${y}: ${checkResult}`;
}
return undefined;
```

I'm also planning to do some cleanup on the checkBuffer/etc helpers this week, so I'll try to remember to revisit this one once I figure out what I'm doing.